File: p2/p256/MinCost.java
package leetcode.p2.p256;

public class MinCost {
    /**
     * DP
     * Time complexity: O(n)
     * Space complexity: O(1)
     */
    public int minCost(int[][] costs) {
        if (costs.length == 0) {
            return 0;
        }
        for (int i = 1; i < costs.length; i++) {
            costs[i][0] = Math.min(costs[i - 1][1], costs[i - 1][2]) + costs[i][0];
            costs[i][1] = Math.min(costs[i - 1][0], costs[i - 1][2]) + costs[i][1];
            costs[i][2] = Math.min(costs[i - 1][1], costs[i - 1][0]) + costs[i][2];
        }
        return Math.min(costs[costs.length - 1][0], Math.min(costs[costs.length - 1][1], costs[costs.length - 1][2]));
    }
}

File: p2/p256/min_cost.py
from typing import List


class Solution:
    """
    Oct 30, 2024 13:33
    DP
    Time Complexity: O(n)
    Space Complexity: O(n)
    """

    def minCost(self, costs: List[List[int]]) -> int:
        m = len(costs)
        n = len(costs[0])
        dp = [[0] * n for _ in range(m)]
        dp[0] = costs[0]
        for i in range(1, m, 1):
            for j in range(n):
                dp[i][j] = float('inf')
                for k in range(n):
                    if j != k:
                        dp[i][j] = min(dp[i][j], dp[i - 1][k] + costs[i][j])

        return min(dp[-1])
